ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx

```diff
@@ -390,7 +397,56 @@ export class WorkflowDag extends React.Component<WorkflowDagProps, WorkflowDagRe
         return {nodes, edges};
     }
 
-    private layoutGraph(nodes: string[], edges: Edge[]) {
+    private layoutGraphPretty(nodes: string[], edges: Edge[]) {
+        const graph = new dagre.graphlib.Graph();
+
+        graph.setGraph({
+            edgesep: 20 / this.scale,
+            nodesep: 50 / this.scale,
+            rankdir: this.state.horizontal ? 'LR' : 'TB',
+            ranksep: 50 / this.scale
+        });
+
+        graph.setDefaultEdgeLabel(() => ({}));
+
+        nodes.forEach(node => {
+            graph.setNode(node, {label: node, width: this.nodeSize, height: this.nodeSize});
+        });
+        edges.forEach(edge => {
+            if (edge.v && edge.w) {
+                graph.setEdge(edge.v, edge.w);
+            }
+        });
+        dagre.layout(graph);
+
+        const size = this.getGraphSize(graph.nodes().map((id: string) => graph.node(id)));
+        this.graph = {
+            width: size.width,
+            height: size.height,
+            nodes: new Map<string, {x: number; y: number}>(),
+            edges: []
+        };
+
+        graph.nodes().map((id: string) => {
```

Comment:
I think this should probably be a singe foreach rather than 2 map

Reply:
This needs to happen after `dagre.layout(graph);` is run so that we get x and y coordinates.